base/file.jl

```diff
@@ -578,10 +578,14 @@ else # !windows
 # Obtain a temporary filename.
 function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
     isdir(parent) || throw(ArgumentError("$(repr(parent)) is not a directory"))
-    p = ccall(:tempnam, Cstring, (Cstring, Cstring), parent, temp_prefix)
-    systemerror(:tempnam, p == C_NULL)
-    s = unsafe_string(p)
-    Libc.free(p)
+    # If TEMPDIR is set tempname will ignore `parent` as an argument (#38873)
+    s = withenv("TMPDIR" => nothing) do
```

Comment:
shouldn't this be conditional on `parent != tempdir()` ?

Reply:
See https://github.com/JuliaLang/julia/pull/38879#issuecomment-744735667

IMO it is fine to do this unconditional. We unset `TMPDIR` here after checking it with `tempdir`, so that the libc implementation will not override `parent` with the value of `TMPDIR`

Reply:
I have a mortal fear of messing with environment variables now.  I'm pretty sure that with this change, we can crash Julia on glibc if we call `tempanme()` in parallel.

I think the only way we can avoid threading issues here is to un-set `TMPDIR` at startup and emulate it. :(

Reply:
At this point I am almost thinking it would be better to avoid the libc implementation of tempname altogether.

Reply:
Should `withenv` have a mutex lock?

Reply:
Avoiding `libc` is probably better. Anything that requires modifying the environment is going to be broken.

Reply:
The [musl implementation](https://git.musl-libc.org/cgit/musl/tree/src/temp) seems pretty straightforward, and is MIT-licensed if we want to copy it.